### PR TITLE
sam4l: Low power I2C

### PR DESCRIPTION
**Pull request by ppannuto**
### Pull Request Overview

This allows the core to enter deep sleep while acting as an I2C slave and still receive messages.

Pulling out this from the middle commit as it deserves attention:

> Critically, this allows the core to sleep while the PDCA clock (Peripheral DMA
> Controller) is active, which according to p320:
>
>  ##### 16.4.1 Power Management
>      If the CPU enters a sleep mode that disables the PDCA clocks, the PDCA
>      will stop functioning and resume operation after the system wakes up from
>      sleep mode.
>
> I think this is the behavior we actually want. If a peripheral has an
> outstanding DMA transaction, it knows about it, and the _peripheral_ can
> inhibit low-power mode. If the peripheral doesn't care, then let the core sleep.

### Testing Strategy

This pull request was tested by programming two hails with the included ping-pong test app and measuring the power draw during operation.

### TODO or Help Wanted

This pull request depends on #760 for correct operation of the I2C peripheral.

### Documentation Updated

- [x] Kernel: The relevant files in `/docs` have been updated or no updates are required.
- [x] Userland: The application README has been added, updated, or no updates are required.

### Formatting

- [x] `make formatall` has been run.

### CODE REVIEW

**Comment by ppannuto** [comment]
This is now rebased on latest master and I2C still works. Should be good to go.

File: chips/sam4l/src/pm.rs [hunk 1]
```diff
@@ -417,16 +417,16 @@ macro_rules! get_clock {
 //
 // This is identical to the reset value of the HSBMASK except it allows the
 // PicoCache RAM clock to be on as well.
-const DEEP_SLEEP_HSBMASK: u32 = 0x1e6;
+const DEEP_SLEEP_HSBMASK: u32 = 0x1e7;
```
Comment (brghena) [inline chips/sam4l/src/pm.rs]: Add PDCA to above comment

File: chips/sam4l/src/pm.rs [hunk 2]
```diff
@@ -417,16 +417,16 @@ macro_rules! get_clock {
...
-const DEEP_SLEEP_PBAMASK: u32 = 0x0;
+// Allow TWIS clocks to be active as they can wake the core from deep sleep.
+const DEEP_SLEEP_PBAMASK: u32 = 0x6000a0;
 
 // FLASHCALW and HRAMC1 clocks allowed
 //
 // This is identical to the reset value of the PBBMASK except it allows the
 // flash's HRAMC1 clock as well.
-const DEEP_SLEEP_PBBMASK: u32 = 0x3;
+const DEEP_SLEEP_PBBMASK: u32 = 0xb;
```
Comment (brghena) [inline chips/sam4l/src/pm.rs]: Add PDCA to above comment

File: chips/sam4l/src/pm.rs [hunk 3]
```diff
@@ -417,16 +417,16 @@ macro_rules! get_clock {
 //
 // This is identical to the reset value of the HSBMASK except it allows the
 // PicoCache RAM clock to be on as well.
-const DEEP_SLEEP_HSBMASK: u32 = 0x1e6;
+const DEEP_SLEEP_HSBMASK: u32 = 0x1e7;
 
-// No clocks allowed on PBA
-const DEEP_SLEEP_PBAMASK: u32 = 0x0;
+// Allow TWIS clocks to be active as they can wake the core from deep sleep.
+const DEEP_SLEEP_PBAMASK: u32 = 0x6000a0;
```
Comment (brghena) [inline chips/sam4l/src/pm.rs]: Why does this allow TWIM2 and TWIM3 to be active too?

**Comment by ppannuto** [comment]
Instead of just changing comments, I updated the `pm` module to the new registers interface.

To be clear, this is only a partial update. It adds all of the _definitions_ for the PM registers, but it only uses them for the sleep clocks. A separate, future PR will be needed to convert PM to actually use the new register interface, but I didn't want to conflate that into this PR.

**Comment by alevy** [comment]
This is even better.

**Pull request merged**